Refactored ForeachFunctors.cuh

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46664 Move device guard from MultiTensorApply.cuh
* #46663 Remove .pyi from /_multi_tensor
* #46662 Renamed a TensorListMetaData property. Cleaned up a test
* #46661 Optimize arguments checks
* **#46660 Refactored ForeachFunctors.cuh**

Differential Revision: [D24453345](https://our.internmc.facebook.com/intern/diff/D24453345)

Commit:
Refactored ForeachFunctors.cuh

[ghstack-poisoned]

Commit:
Update on "Refactored ForeachFunctors.cuh"




[ghstack-poisoned]

cpuhrsch reviewed (changes requested):
The ROCM failures look legit and related https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.8-py3.6-test2/2268/console 

```
19:08:18 ======================================================================
19:08:18 FAIL: test_adam (__main__.TestOptim)
19:08:18 ----------------------------------------------------------------------
19:08:18 Traceback (most recent call last):
19:08:18   File "test_optim.py", line 386, in test_adam
19:08:18     lambda weight, bias: optimizer([weight, bias], lr=1e-3)
19:08:18   File "test_optim.py", line 208, in _test_basic_cases
19:08:18     constructor
19:08:18   File "test_optim.py", line 192, in _test_state_dict
19:08:18     self.assertEqual(weight, weight_cuda)
19:08:18   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1144, in assertEqual
19:08:18     self.assertTrue(result, msg=msg)
19:08:18 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 50 element(s) (out of 50) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 997835.0437917262 (0.143708273768425 vs. 997835.1875), which occurred at index (6, 4).
```

I'm rerunning just to make sure.

Commit:
Update on "Refactored ForeachFunctors.cuh"


Differential Revision: [D24453345](https://our.internmc.facebook.com/intern/diff/D24453345)

[ghstack-poisoned]

Commit:
Update on "Refactored ForeachFunctors.cuh"


Differential Revision: [D24453345](https://our.internmc.facebook.com/intern/diff/D24453345)

[ghstack-poisoned]

izdeby commented:
@cpuhrsch, ROCM is fixed now. Rebased.

ngimel commented on aten/src/ATen/native/cuda/AmpKernels.cu:
```diff
@@ -156,7 +156,10 @@ void _amp_foreach_non_finite_check_and_unscale_cuda_(TensorList scaled_grads,
       multi_tensor_apply<1>(tensor_lists,
-                            UnaryOpFunctor_<scalar_t>(),
+                            UnaryOpFunctor<scalar_t, 
+                                           /* depth */ 1,
```
nit: it looks like everywhere `depth` is equal to `multi_tensor_apply` template argument (and moreover, things would break if it isn't), so maybe use constexpr variable there?

ngimel reviewed (commented):
Thank you, this is a much needed refactor!
